src/shell/server.cpp

```diff
@@ -146,4 +168,172 @@ json server::handle_check(json const &) {
+    if (auto it = is_essentially_atomic(env, d.get_name())) {
+        std::string it_str = it->to_string();
+        // if pattern "perfectly" matches beginning of declaration name, we just display d on the top of the list
+        if (it_str.compare(0, pattern.size(), pattern) == 0)
+            return it;
+    } else {
+        std::string d_str = d.get_name().to_string();
+        if (d_str.compare(0, pattern.size(), pattern) == 0)
+            return optional<name>(d.get_name());
+    }
+    return optional<name>();
+}
+
+json server::serialize_decl(name const & short_name, name const & long_name, environment const & env, options const & o) {
+    declaration const & d = env.get(long_name);
+    type_context tc(env);
+    io_state_stream out   = regular(env, m_ios, tc).update_options(o);
+    expr type = d.get_type();
+    if (LEAN_COMPLETE_CONSUME_IMPLICIT) {
+        while (true) {
+            if (!is_pi(type))
+                break;
+            if (!binding_info(type).is_implicit() && !binding_info(type).is_inst_implicit())
+                break;
+            std::string q("?");
+            q += binding_name(type).to_string();
+            expr m = mk_constant(name(q.c_str()));
+            type   = instantiate(binding_body(type), m);
+        }
+    }
+    json completion;
+    completion["text"] = short_name.to_string();
+    std::ostringstream ss;
+    ss << mk_pair(flatten(out.get_formatter()(type)), o);
+    completion["type"] = ss.str();
+    return completion;
+}
+
+json server::serialize_decl(name const & d, environment const & env, options const & o) {
+    // using namespace override resolution rule
+    list<name> const & ns_list = get_namespaces(env);
+    for (name const & ns : ns_list) {
+        name new_d = d.replace_prefix(ns, name());
+        if (new_d != d &&
+            !new_d.is_anonymous() &&
+            (!new_d.is_atomic() || !is_protected(env, d))) {
+            return serialize_decl(new_d, d, env, o);
+        }
+    }
+    // if the alias is unique use it
+    if (auto it = is_uniquely_aliased(env, d)) {
+        return serialize_decl(*it, d, env, o);
+    } else {
+        return serialize_decl(d, d, env, o);
+    }
+}
+
+json server::handle_complete(json const & req) {
+    std::string pattern = req["pattern"];
+    unsigned line = req["line"];
+    std::vector<json> completions;
+
+    if (snapshot const * snap = get_closest_snapshot(line)) {
+        environment const & env = snap->m_env;
+        options const & opts = snap->m_options;
+        unsigned max_results = get_auto_completion_max_results(opts);
+        unsigned max_errors = get_fuzzy_match_max_errors(pattern.size());
+        std::vector<pair<name, name>> exact_matches;
+        std::vector<pair<std::string, name>> selected;
+        bitap_fuzzy_search matcher(pattern, max_errors);
+        env.for_each_declaration([&](declaration const & d) {
+            /*if (is_projection(env, d.get_name()))
+                return;*/
```

Comment:
> It would be nice to only hide inherited projections
I'm not sure it would cover all use-cases. Suppose we have the following two structures:
```
structure A := (x : nat)
structure B extends A := (y : nat)
```
They are not classes. If we have `b : B`, we probably want to use `B.x` instead of `A.x`. To use `A.x` we will first have to coerce `b` into `A`.